sql-plugin/src/main/scala/com/nvidia/spark/rapids/ParquetSchemaUtils.scala

```diff
@@ -0,0 +1,574 @@
+      parquetType.getFields.asScala.map(f => f.getName -> f).toMap
+    lazy val caseInsensitiveParquetFieldMap =
+      parquetType.getFields.asScala.groupBy(_.getName.toLowerCase(Locale.ROOT))
+    lazy val idToParquetFieldMap =
+      parquetType.getFields.asScala.filter(_.getId != null).groupBy(f => f.getId.intValue())
+
+    def updateField(oldField: StructField, p: Type): StructField = {
+      val newSparkType = clipSparkType(oldField.dataType, p, caseSensitive, useFieldId)
+      oldField.copy(dataType = newSparkType)
+    }
+
+    def matchCaseSensitiveField(f: StructField): Option[StructField] = {
+      caseSensitiveParquetFieldMap
+          .get(f.name)
+          .map(updateField(f, _))
+    }
+
+    def matchCaseInsensitiveField(f: StructField): Option[StructField] = {
+      // Do case-insensitive resolution only if in case-insensitive mode
+      caseInsensitiveParquetFieldMap
+          .get(f.name.toLowerCase(Locale.ROOT))
+          .map { parquetTypes =>
+            if (parquetTypes.size > 1) {
+              // Need to fail if there is ambiguity, i.e. more than one field is matched
+              val parquetTypesString = parquetTypes.map(_.getName).mkString("[", ", ", "]")
+              throw RapidsErrorUtils.foundDuplicateFieldInCaseInsensitiveModeError(
+                f.name, parquetTypesString)
+            } else {
+              updateField(f, parquetTypes.head)
+            }
+          }
+    }
+
+    def matchIdField(f: StructField): Option[StructField] = {
+      val fieldId = ParquetSchemaClipShims.getFieldId(f)
+      idToParquetFieldMap
+          .get(fieldId)
+          .map { parquetTypes =>
+            if (parquetTypes.size > 1) {
+              // Need to fail if there is ambiguity, i.e. more than one field is matched
+              val parquetTypesString = parquetTypes.map(_.getName).mkString("[", ", ", "]")
+              throw new RuntimeException(
+                s"""
+                   |Found duplicate field(s) "$fieldId": $parquetTypesString
+                   |in case-insensitive mode
+                 """.stripMargin.replaceAll("\n", " "))
+            } else {
+              updateField(f, parquetTypes.head)
+            }
+          }
+    }
+
+    val shouldMatchById = useFieldId && ParquetSchemaClipShims.hasFieldIds(sparkType)
+    val updatedFields = sparkType.flatMap { f =>
+      if (shouldMatchById && ParquetSchemaClipShims.hasFieldId(f)) {
+        matchIdField(f)
+      } else if (caseSensitive) {
+        matchCaseSensitiveField(f)
+      } else {
+        matchCaseInsensitiveField(f)
+      }
+    }
+    StructType(updatedFields)
+  }
+
+  def evolveSchemaIfNeededAndClose(
+      table: Table,
+      fileSchema: MessageType,
+      sparkSchema: StructType,
+      caseSensitive: Boolean,
+      useFieldId: Boolean): Table = {
+    val fileSparkSchema = closeOnExcept(table) { _ =>
+      clipSparkSchema(sparkSchema, fileSchema, caseSensitive, useFieldId)
```

Comment:
I am wondering whether this `clipSparkSchema` can be removed after

- In `clipParquetSchema`, building the `clipedSchema` (the schema of the memory Parquet file) with each field is generated with the name from the corresponding field (matched by field id or column name) in `catalystSchema` (the read schema) but others still from the field in `parquetSchema` (the file schema).

   Assuming there is a field **_(name: "c1", id: 2, type: int)_** in `parquetSchema`, and a field **_(name: "cc", id: 2, type: long)_** in `catalystSchema`. They will match by id if `useFieldId` is true. Then the final output field will be _**(name: "cc", id: 2, type: int)**_.
```
        (name: "c1", id: 2, type: int)    // In fileSchema 
    +   (name: "cc", id: 2, type: long)  // In readSchema
    ->  (name: "cc", id: 2, type: int)
```
  Then inside the coming `SchemaUtils.evolveSchemaIfNeededAndClose`, the name comparsion will be enough for Parquet. 
  [We are doing this in ORC](https://github.com/NVIDIA/spark-rapids/blob/branch-22.08/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala#L1214). So the name only comparsion works for ORC even ORC also supports [a positional way to match columns](https://github.com/NVIDIA/spark-rapids/blob/branch-22.08/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala#L1192), which is similar to the Parquet field id.
- Building the column names for cuDF decoding from this `clipedSchema` instead of the `toCudfColumnNames` method. Since with the `SchemaUtils.evolveSchemaIfNeededAndClose`, the input table is not required to have the same column order with the reader schema any more.

Reply:
I don't think this will work as stated.  I see the following problems, please correct me if I'm missing something.

The `MessageType` that comes out of the new `checkSchemaCompat` may not match the column paths that are encoded in the `BlockMetaData` corresponding to the data blocks in the file.  When we emit the in-memory Parquet file using the new `MessageType` that is using Spark's names instead of the original file's column names, the libcudf Parquet reader won't be able to find the the column data for some columns (or worse, will use the _wrong_ column data for the corner case where column names were swapped).  Writing a footer schema that doesn't match the column data blocks is essentially the root cause of #6005.  The footer schema says there's a column for which there is no corresponding column data that matches that column.  We might be able to fixup the `BlockMetaData` for the new column name mapping, but I don't know offhand how hard this would be.

> Then it will become the table schema inside this SchemaUtils.evolveSchemaIfNeededAndClose function.

That method doesn't take a `MessageType`, it takes only Spark schemas since it isn't specific to Parquet.  We'd still need a way to convert a `MessageType` into a `StructType`, which would end up with code very similar to `clipSparkSchema`.

Reply:
OK, I missed that. So we also need to rebuild the block meta if using this "new" schema.